opentelemetry/src/context.rs

```diff
@@ -381,6 +387,87 @@ impl Hasher for IdHasher {
     }
 }
 
+/// A stack for keeping track of the [`Context`] instances that have been attached
+/// to a thread.
+///
+/// The stack allows for popping of contexts by position, which is used to do out
+/// of order dropping of [`ContextGuard`] instances. Only when the top of the
+/// stack is popped, the topmost [`Context`] is actually restored.
+///
+/// The stack relies on the fact that it is thread local and that the
+/// [`ContextGuard`] instances that are constructed using it can't be shared with
+/// other threads.
+struct ContextStack {
+    /// This is the current [`Context`] that is active on this thread, and the top
+    /// of the [`ContextStack`]. It is always present, and if the `stack` is empty
+    /// it's an empty [`Context`].
+    ///
+    /// Having this here allows for fast access to the current [`Context`].
+    current_cx: Context,
+    /// A `stack` of the other contexts that have been attached to the thread.
+    stack: Vec<Option<Context>>,
+    /// Ensure this type is !Send as it relies on thread locals
+    _marker: PhantomData<*const ()>,
+}
+
+impl ContextStack {
+    #[inline(always)]
+    fn push(&mut self, cx: Context) -> usize {
+        // The next id is the length of the `stack`, plus one since we have the
+        // top of the [`ContextStack`] as the `current_cx`.
+        let next_id = self.stack.len() + 1;
+        let current_cx = std::mem::replace(&mut self.current_cx, cx);
+        self.stack.push(Some(current_cx));
+        next_id
+    }
+
+    #[inline(always)]
+    fn pop_id(&mut self, pos: usize) {
```

Comment:
I believe this isn't reentrant - there's some checking of IDs and then _subsequent action_ ™️  mutating the context and the current_cx pointer. 

I believe this doesn't matter, because this can never be called more than once concurrently in the same thread. Is this 100% the case? Is there some fun edge case here?

Reply:
It'd be good to get coverage on the two bounds checks in here too

Reply:
Yes, more tests should be added. Also, it is safe. Or at least that's the promise of Rust since we can't share or create the `ContextGuard` instances in any other way.
```
/// The stack relies on the fact that it is thread local and that the
/// [`ContextGuard`] instances that are constructed using it can't be shared with
/// other threads.
```